### ISSUE TITLE
ES Modules for Tree Shaking with Webpack2 and Rollup

### ISSUE THREAD

**Issue by schtauffen** [ramda/ramda#2180]
I would like to have an es-modules version of ramda which can be properly tree-shaken by Webpack2 and Rollup. 

Is this in the works? 
Pull request okay?

Otherwise I will be publishing an unofficial wrapper which does so.

**Comment by kedashoe** [comment]
Yes, see https://github.com/ramda/ramda/issues/1968 and https://github.com/ramda/ramda/pull/1977. I think the plan is to release 1.0 in the near future and then this will be one of the first things we tackle after that.

**Comment by olsonpm** [comment]
@schtauffen - the difficulty is that you can't just turn a cjs library into es modules and call it good.  You should spend the time to try it with ramda, check the tree-shaken output, and learn why it doesn't work.  Or you could read through #1968, but I suggest the former.

**Comment by schtauffen** [comment]
@olsonpm I have spent time looking at this issue and have decided that `babel-plugin-ramda` is sufficient for me at this time. Still hoping it won't require a plugin in the future :)

**Comment by olsonpm** [comment]
Yeah - unfortunately the root issue is the complexity in automatic tree-shaking.  I think static analysis in javascript is just incapable of performing such a task without tons of edge cases and thus brittle, buggy code.  I may be wrong of course, but that's what it seems from the tree-shaking issues I've seen in webpack and rollup.

**Comment by ivan-kleshnin** [comment]
Does anyone have a working demo with Ramda + Webpack + treeshaking?
Webpack seems to have a broken treeshaking: https://github.com/webpack/webpack/issues/2867

**Comment by olsonpm** [comment]
the only way you'll get the desired result is to minify it via uglify and enable the pure function dead code elimination option.  just look at uglifyjs readme and search "pure"

**Comment by kedashoe** [comment]
@ivan-kleshnin you can see some comments about building ramda with webpack here: https://github.com/ramda/ramda/pull/2254#issuecomment-328061077

**Comment by olsonpm** [comment]
it would be awesome if tree-shaking had a ton of disclaimers everywhere it was mentioned to lower everyone's expectations from the magic they initially hope for.  The blog posts and hype around it at the beginning really screwed up the general perception of how it works.

**Comment by ivan-kleshnin** [comment]
> The blog posts and hype around it at the beginning really screwed up the general perception of how it works.

👍 

@olsonpm tree-shaking seems to work in Webpack **only with production flag**. Guess how often this "tiny" fact is mentioned? Right – it's not mentioned in docs at all... 

I believe the code analysis is very expensive but the real news is that it won't speed things up in development. This hype is really toxic.

**Comment by olsonpm** [comment]
kudos for successfully fighting your way through webpack.  i'm going to be doing the same myself in a few weeks and am not looking forward to it.

not that the devs behind the project are bad - they've all contributed much more than i have and i admire them a ton.  But man are those configurations daunting and a battle to get through.

**Comment by scabbiaza** [comment]
After many setup experiments we discovered that Babel `env` plugin breaks treeshaking. The problem is this plugin is recommended as a go-to solution in Babel docs. Without this plugin both UglifyJS webpack plugin or `-p` flag enable tree-shaking. This feature does not provide impressive results with Ramda, btw. because of complex interdependncies between modules. For example, an identity function pulls all the currying machinery 😱 .

**Comment by olsonpm** [comment]
> why curry a unary function
because of behavior that i don't think should exist:

`identity()('hey there') === 'hey there'`

thanks for the tidbits btw

**Comment by buzzdecafe** [comment]
I am also opposed to currying unary functions. IIRC, tt is only there to provide a consistent experience with placeholder, which I also think should be removed. Another one for the post-1.0 world, i guess

**Comment by olsonpm** [comment]
yeah, i think the main issue is determing appropriate behavior for calling a curried function with no arguments.  personally i believe an error should be thrown, but mostly because i haven't come across a case where calling a curried function with no arguments makes sense

**Comment by buzzdecafe** [comment]
100% agree, that is what we used to do.

**Comment by ivan-kleshnin** [comment]
```
curry(fn)() <=> fn() <=> fn(undefined)
```

so

```
curry(id)() <=> id() <=> id(undefined) <=> undefined (and not id)
```

**Comment by CrossEye** [comment]
> `id() <=> id(undefined)`

I don't buy this step.  It took a number of people a lot of patient explanation to demonstrate to me that we can't treat `undefined` this way, as reasonable as it seems.

Thus, we couldn't have `Maybe.from(undefined)` yielding `Nothing()`, it had to yield `Just(undefined)`.  I'm still not thrilled with that, but I understand that it's necessary for various laws to pass.

I feel similarly about calling a curried function with no arguments: there seems to be a law lurking that if `length(a) + length(b) <= n` where `curriedFn` has length `n` and `a` and `b` are argument lists, then calling `curriedFn` with `a` and calling the result with `b` should be the same as calling it with those two lists concatenated.  I simply don't see a reason for an exception when `length(a)` is 0.  But I've never really tried to formalize this feeling.

**Comment by olsonpm** [comment]
My perspective is less from a formalized standpoint and more an end-user's intuition.

I don't think anyone would write `curriedFn()` and intend that.  If they do, I want to understand their use-case because I can't imagine it being anything more than academic.

Maybe formally it makes sense to do one thing over another, but I think you should instead be focused on developer's intent when creating an api, because not matching intent is what produces bug.

**Comment by ivan-kleshnin** [comment]
@olsonpm the problem is calling a function with no arguments is equivalent to calling it with `null` or `undefined` from an *academic* point of view (category theory). But maybe that goes for normal functions, not variadic... JS is insanely complex, a nightmare to describe in math terms.

**Comment by olsonpm** [comment]
in js, calling a function with no arguments is very different from calling with a single null or undefined value, both semantically and functionally (functionally meaning there are noticeable differences in the function call itself, the most obvious being arguments.length)

**Comment by CrossEye** [comment]
@ivan-kleshnin: My point was that `null` and `undefined` are _values_, so calling with them is in fact very different from calling with no arguments.  We can't assume an arbitrary function called with no arguments will have identical behavior to one called with `undefined`.  It's true also for sparse arrays versus ones with explicit values: there are different behaviors, so the values cannot be considered the same.

@olsonpm: My notion is still vague and intuitive rather than actually formal.  But I'm betting that if I tried, I could find a way in which someone could accidentally invoke `curriedFn` with an empty list of arguments in such a way that if we simply returned the equivalent function, the system would continue to work.  By adding an exception when that happens, we would force this user to add error checking, and make it harder for her compose functions.

But part of this is that I simply find exceptions to be mostly horrible, and will do as much as possible to avoid them.  But that seems the only other possibility for `curriedFn()`.
